File: client/src/api/atletaService.js
/**
 * Servicio para gestión de Atletas
 * Conecta con el endpoint /api/v1/atletas del backend
 */

import apiClient from './apiClient'
import { ENDPOINTS } from '../config/api.config'

const AtletaService = {
  /**
   * Obtener todos los atletas
   * @param {Object} params - Parámetros de filtro y paginación
   */
  getAll: async (params = {}) => {
    const response = await apiClient.get(ENDPOINTS.ATLETAS, { params })
    return response.data
  },

  /**
   * Obtener un atleta por ID
   * @param {number} id - ID del atleta
   */
  getById: async (id) => {
    const response = await apiClient.get(`${ENDPOINTS.ATLETAS}${id}/`)
    return response.data
  },

  /**
   * Crear un nuevo atleta
   * @param {Object} data - Datos del atleta
   */
  create: async (data) => {
    const response = await apiClient.post(ENDPOINTS.ATLETAS, data)
    return response.data
  },

  /**
   * Actualizar un atleta existente
   * @param {number} id - ID del atleta
   * @param {Object} data - Datos a actualizar
   */
  update: async (id, data) => {
    const response = await apiClient.put(`${ENDPOINTS.ATLETAS}${id}/`, data)
    return response.data
  },

  /**
   * Eliminar un atleta
   * @param {number} id - ID del atleta
   */
  delete: async (id) => {
    const response = await apiClient.delete(`${ENDPOINTS.ATLETAS}${id}/`)
    return response.data
  },

  /**
   * Buscar atletas por criterios
   * @param {Object} criteria - Criterios de búsqueda
   */
  search: async (criteria) => {
    const response = await apiClient.get(`${ENDPOINTS.ATLETAS}/buscar`, { params: criteria })
    return response.data
  },
}

export default AtletaService

File: server/basketball/services/atleta_service.py
"""
Servicio para la gestión de Atletas
"""

import logging
from typing import List, Optional, Dict, Any
from django.core.exceptions import ValidationError
from ..dao.atleta_dao import AtletaDAO
from ..models import Atleta

logger = logging.getLogger(__name__)


class AtletaService:
    """Servicio que encapsula la lógica de negocio para Atletas"""
    
    def __init__(self):
        self.dao = AtletaDAO()
    
    def create_atleta(self, data: Dict[str, Any]) -> Atleta:
        """
        Crea un nuevo atleta.
        
        Args:
            data: Diccionario con los datos del atleta
            
        Returns:
            Atleta: Instancia creada
            
        Raises:
            ValidationError: Si los datos no son válidos
        """
        dni = data.get('dni')
        if not dni:
            raise ValidationError("El DNI es obligatorio")
            
        if self.dao.exists_dni(dni):
            raise ValidationError(f"Ya existe un atleta con el DNI {dni}")
            
        return self.dao.create(**data)
    
    def update_atleta(self, pk: int, data: Dict[str, Any]) -> Optional[Atleta]:
        """
        Actualiza un atleta existente.
        
        Args:
            pk: ID del atleta
            data: Datos a actualizar
            
        Returns:
            Atleta | None: Instancia actualizada
        """
        if 'dni' in data:
            if self.dao.exists_dni(data['dni'], exclude_id=pk):
                raise ValidationError(f"Ya existe otro atleta con el DNI {data['dni']}")
        
        return self.dao.update(pk, **data)
    
    def delete_atleta(self, pk: int) -> bool:
        """
        Elimina (soft delete) un atleta.
        
        Args:
            pk: ID del atleta
            
        Returns:
            bool: True si se eliminó correctamente
        """
        return self.dao.soft_delete(pk)
    
    def get_atleta_by_id(self, pk: int) -> Optional[Atleta]:
        """Obtiene un atleta por ID"""
        return self.dao.get_by_id(pk)
    
    def get_all_atletas(self) -> List[Atleta]:
        """Obtiene todos los atletas activos"""
        return list(self.dao.get_activos())
    
    def search_atletas(self, search_term: str) -> List[Atleta]:
        """Busca atletas por término"""
        return list(self.dao.search_atletas(search_term))
    
    def asignar_grupo(self, atleta_id: int, grupo_id: int) -> bool:
        """Asigna un atleta a un grupo"""
        return self.dao.asignar_grupo(atleta_id, grupo_id)
    
    def remover_grupo(self, atleta_id: int, grupo_id: int) -> bool:
        """Remueve un atleta de un grupo"""
        return self.dao.remover_grupo(atleta_id, grupo_id)
